bokeh/io/saving.py

```diff
@@ -106,46 +116,47 @@ def _get_save_args(state, filename, resources, title):
 
     return filename, resources, title
 
-def _get_save_filename(state, filename):
+def _get_save_filename(state: State, filename: Optional[str]) -> Tuple[str, bool]:
     if filename is not None:
         return filename, False
 
     if state.file and not settings.ignore_filename():
-        return state.file['filename'], False
+        return state.file.filename, False
```

Comment:
@bryevdv, note that replacing `dict` with dataclasses is a breaking change. It's easily possible to make dataclasses backwards compatible with regards to this, with the side effect that we will have to import our own dataclass derivation instead of the built-in one.

Reply:
I don't think users should be using state objects directly, I have also never encountered a single support question about them. I would even say we should make all of this explicitly private. In any case I am not worried about this change, but put a note in the release notes for good measure.

Reply:
I added a release note. However, I was speaking in general regarding interim support for dataclasses with `dict`-style indexing for the sake of backwards compatibility, because I intend to proceed with similar changes across the codebase (e.g. replace `dict(field=str)` with an appropriate dataclass, though that will require additional `accepts` on relevant properties).

Reply:
For sure, there are other situations like that that will need a compat interface, we can take them on a case-by-case basis.